// ts-eventemitter/dist/ts-eventemitter.js
"use strict";
var events_1 = require('events');
var TsEventEmitterImpl = (function () {
    function TsEventEmitterImpl() {
        this.eventEmitter = new events_1.EventEmitter();
    }
    TsEventEmitterImpl.prototype.event = function (name) {
        return new EventImpl(this, this.eventEmitter, name);
    };
    TsEventEmitterImpl.prototype.removeAllListeners = function () {
        this.eventEmitter.removeAllListeners();
    };
    TsEventEmitterImpl.prototype.setMaxListeners = function (n) {
        this.eventEmitter.setMaxListeners(n);
    };
    return TsEventEmitterImpl;
}());
var EventImpl = (function () {
    function EventImpl(tsEventEmitter, eventEmitter, name) {
        this.tsEventEmitter = tsEventEmitter;
        this.eventEmitter = eventEmitter;
        this.name = name;
    }
    EventImpl.prototype.on = function (listener) {
        this.eventEmitter.on(this.name, listener);
        return this.tsEventEmitter;
    };
    EventImpl.prototype.addListener = function (listener) {
        this.eventEmitter.addListener(this.name, listener);
        return this.tsEventEmitter;
    };
    EventImpl.prototype.off = function (listener) {
        return this.removeListener(listener);
    };
    EventImpl.prototype.removeListener = function (listener) {
        this.eventEmitter.removeListener(this.name, listener);
        return this.tsEventEmitter;
    };
    EventImpl.prototype.removeAllListeners = function () {
        this.eventEmitter.removeAllListeners(this.name);
        return this.tsEventEmitter;
    };
    EventImpl.prototype.listeners = function () {
        return this.eventEmitter.listeners(this.name);
    };
    EventImpl.prototype.once = function (listener) {
        this.eventEmitter.once(this.name, listener);
        return this.tsEventEmitter;
    };
    EventImpl.prototype.emit = function () {
        var args = [];
        for (var _i = 0; _i < arguments.length; _i++) {
            args[_i - 0] = arguments[_i];
        }
        args.unshift(this.name);
        return this.eventEmitter.emit.apply(this.eventEmitter, args);
    };
    return EventImpl;
}());
var TsEventEmitter;
(function (TsEventEmitter) {
    function create() {
        return new TsEventEmitterImpl();
    }
    TsEventEmitter.create = create;
})(TsEventEmitter || (TsEventEmitter = {}));
exports.__esModule = true;
exports["default"] = TsEventEmitter;
